### PR TITLE
Refactor MCM: Minimize disk operations

### PR DESCRIPTION
**Pull request by herbertgithubaccount**
This originally started off as part of  #532, but has been split off into a separate PR.

This summary was taken from #532 :

This is a pretty sizeable PR that makes a couple large changes to the MCM. The main changes are:

### Description
Before this PR, there was a complicated block of code that traversed the `mcm/components` and `mcm/variables` directories in order to generate the `create<Component|Variable>` methods. This happened as follows:
- Define an `__index(tbl, key)` metamethod that checks if `key` starts with `"create"`.
- If `key:startswith("create")`, then scan the `mcm/components` and `mcm/variables` directories until we find the matching file. Then `require` that file, and do some object creation logic.
- If `not key:startswith("create")`, then lookup the key in the parent class.

This same block of code was repeated in the following files and functions:
- `Component:getComponent`
- `Category.__index`
- `Template.__index`
- `mwse.mcm.__index`

This PR makes the following changes:
- Add (hidden) `mwse.mcm.getComponentClass` and `mwse.mcm.getVariableClass` functions that take in a `className` and return the corresponding component.
   - These functions will cache their return results, so they only involve I/O operations once per `className`.
   - Additionally, the `mwse/components` and `mwse/variables` directory are scanned only once per launch, and the results are cached prior to  `mwse.mcm.getComponentClass` or `mwse.mcm.getVariableClass` being called.
      - Rudimentary testing has shown this results in a noticeable improvement in performance, with little added complexity.
- Change the `mwse.mcm.__index` metamethod so that:
   - It uses the `getComponentClass` and `getVariableClass` functions.
   - The `create<Component|Variable>` functions are generated whenever accessed.
      - This means that the `__index` metamethod only triggers once per call to `create<Component|Variable>`
- Alter `Category.__index`, and `Template.__index` so that they make use of `mwse.mcm.create<Component|Variable>`
   - Make `Category.__index` and `Tempalte.__index` cached their results so that the `create<Component|Variable>` logic only happens once.
- Alter `Component:getComponent` so that it uses the `mwse.mcm.getComponentClass` and `mwse.mcm.getVariableClass` functions.

### Results/Consequences:
- These changes dramatically reduce code duplication among those four files
- Basic testing shows these changes also result in roughly a `0.4 to 0.5` second improvement to runtime of `event.trigger("modConfigReady")`.
    - Testing was done when Morrowind was installed on an SSD, and was done using both a large mod list and a small mod list.
    - The performance gains do not seem to depend on the size of the installed mod lists:
        - Small mod list - UNOPTIMIZED: `0.5345630645752 seconds`
        - Small mod list - OPTIMIZED:   `0.095428466796875 seconds`
        - Big mod list   - UNOPTIMIZED: `6.2415904998779 seconds`
        - Big mod list   - OPTIMIZED:   `5.7679977416992 seconds`

    - I'm not sure how much these values depend on the hard drive/OS being used.
    - Time taken to load a mod page (e.g. after clicking on a `modNameButton`) was not noticeably different for most mods.

### Other Changes
- The `mcm/mcm.lua` file was changed so that it's no longer a singleton class.
    - Prior to this change, `mcm/mcm.lua` defined a `mcm:new()` method, and then returned a new instance of the `mcm` table.
        - This resulted in the complex `__index` method getting called each time any key was accessed in `mwse.mcm`.
        - It also seemed inconsistent with the implementation of many other global tables.
- The `mcm/classPaths.lua` file is no longer needed, and thus has been marked as `deprecated`.
    - `classPaths.lua` contains hardcoded information about the directory structure of the `components` and `variables` folders, and it was used in each of `Component:getComponent`, `Category.__index`, `Template.__index`, and `mwse.mcm.__index` functions/mathods.
    - Previously, if a subdirectory was not present in `classPaths.lua`, then the corresponding `create<Class>` functions/methods would not be generated.
    - This PR makes it so the `core/mcm/components` and `core/mcm/variables` directories are traversed exactly once using `lfs.walkdir` in the `mcm/mcm.lua` file. The directory structure is no longer utilized in any other `mcm` files.
        - As a result, `classPaths.lua` is no longer necessary.

### CODE REVIEW

File: misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua [hunk 1]
```diff
@@ -23,8 +24,7 @@ function Setting:new(data)
 	if data and data.variable then
 		-- create setting variable
 		t.variable.defaultSetting = t.variable.defaultSetting or t.defaultSetting
-		local typePath = ("mcm.variables." .. t.variable.class)
-		t.variable = require(typePath):new(t.variable)
+		t.variable = fileUtils.getVariableClass(t.variable.class):new(t.variable)
```
Comment (herbertgithubaccount) [inline misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua:27]: This change doesn't have much impact on performance, but it maintains consistency with the other changes. It also helps to make sure `fileUtils` is solely responsible for all logic that associates a Lua file to a given `className`.

File: misc/package/Data Files/MWSE/core/mcm/components/templates/Template.lua [hunk 2]
```diff
@@ -331,34 +331,27 @@ function Template:register()
 end
 
 function Template.__index(tbl, key)
```
Comment (herbertgithubaccount) [inline misc/package/Data Files/MWSE/core/mcm/components/templates/Template.lua:333]: The current implementation of `Template.__index` and `Category.__index` are basically piggybacking off of `mwse.mcm`. This may not be the most optimized approach, but it means that most of the complicated string logic is only done in `mwse/mcm.lua` and `mcm/fileUtils.lua`.

I'd be happy to change the implementations of `Template.__index` and `Category.__index` to something else if desired though.

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+local components = {}
+local variables = {}
+
+---@protected
+---@param className string The name of the class of the `mwseMCMComponent` to search for.
+---@return mwseMCMComponent?
+function fileUtils.getComponentClass(className)
+	local class = components[className]
+	if class then return class end
+	if className == "HyperLink" then
```
Comment (herbertgithubaccount) [inline misc/package/Data Files/MWSE/core/mcm/fileUtils.lua:37]: Some mods wanted to call `createHyperLink` instead of `createHyperlink`, so I hardcoded a safeguard into this function. Same thing happened with `createSidebarPage` instead of `createSideBarPage`.

A more general solution might be preferable though.

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+local componentPaths = {}
+local variablePaths = {}
+
+for filePath, dir, fileName in lfs.walkdir("data files\\mwse\\core\\mcm\\components\\") do
+	-- For example, when adding the path of `mwseMCMPage`:
+	-- The key   will be:  "Page" 
+	--        instead of:  "Page.lua"
+	-- The value will be:  "mcm.components.pages.Page" 
+	-- 		  instead of:  "data files\\mwse\\core\\mcm\\components\\pages\\Page.lua"
+	componentPaths[fileName:sub(1, fileName:len() - 4)] = filePath:sub(1 + prefixLength, filePath:len() - 4):gsub("[\\/]", ".")
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/fileUtils.lua:21]: Now to make this easier to understand, use code to explain code instead of long comments:
- extract `fileName:sub(1, fileName:len() - 4)` into a `function removeLuaExtension(fileName)` function

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+---@return mwseMCMComponent?
+function fileUtils.getComponentClass(className)
+	local class = components[className]
+	if class then return class end
+	if className == "HyperLink" then
+		return fileUtils.getComponentClass("Hyperlink")
+	end
+	if className == "SidebarPage" then
+		return fileUtils.getComponentClass("SideBarPage")
+	end
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/fileUtils.lua:42]: These things look cursed. Have you considered storing the keys of `componentPaths` and `variablePaths` in lowercase?

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	local luaPath = componentPaths[className]
+	if luaPath then
+		class = include(luaPath)
+		if class and type(class) == "table" then
+			class.class = className -- Store it now so we don't have to do this every time.
+			components[className] = class
+			return class
+		end
+	end
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/fileUtils.lua:51]: Reduce the nesting by inverting the if condition:
```lua
if not luaPath then return end
-- ...
```

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	if class then return class end
+	local luaPath = variablePaths[className]
+	if luaPath then
+		class = include(luaPath)
+		if class and type(class) == "table" then
+			class.class = className -- Store it now so we don't have to do it every time.
+			variables[className] = class
+			return class
+		end
+	end
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/fileUtils.lua:68]: Same as above.

File: misc/package/Data Files/MWSE/core/mcm/mcm.lua [hunk 8]
```diff
@@ -133,61 +127,65 @@ end
...
 	end
+	-- Now check if it's a variable.
+	local variableClass = fileUtils.getVariableClass(className)
+	if variableClass then
+		-- Store the function so we don't have to recreate it every time.
+		mcm[key] = function(param1, param2)
+			return variableClass:new(param2 or param1)
+		end
+		return mcm[key]
+	end
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/mcm.lua:187]: Again, you can reduce the nesting level by inverting the if condition.

**Review by C3pa — CHANGES_REQUESTED** [review]
Here are some of my initial thoughts.

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua (hunk 3 above)
Comment (C3pa) [reply misc/package/Data Files/MWSE/core/mcm/fileUtils.lua:37]: > Maybe it's better to store things in the components and variables tables by first calling string.lower on the keys? And then calling string.lower on the className parameter

You've just surpassed me 😁.

About the performance impact, I can't really say. As always, performance should be measured before reasoning. Nevertheless, I'd take this in even if it doesn't end up faster than the master because it reduces code duplication.

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua (hunk 6 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/fileUtils.lua:51]: Great idea! I usually try to do that because I agree that it makes the code much more readable. But it looks like I missed this one. I'll change it now.

File: misc/package/Data Files/MWSE/core/mcm/mcm.lua (hunk 8 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/mcm.lua:187]: In this particular case I ended up not inverting the if condition (in the original PR) because I thought it made the `variableClass` part of the code more symmetric with the `componentClass` part of the function. i.e., the function has two main chunks: `if componentClass then ... end` and then later `if variableClass then ... end`, and the symmetrical `if <class> then ... end` checks helped me mentally break up the function into two parts.

I'll make the change you suggested. I just wanted to mention this so that I could ask about the more general setting so that I know for next time: is it generally better to prioritize reducing the nesting level vs maintaining "symmetry" with prior code blocks?

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua (hunk 3 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/fileUtils.lua]: The performance advice is very sound.

I thought of another solution that doesn't use `string.lower`, although it still feels like it may be a little unstable: I can remove the cursed `if className == ...` checks by adding the following two lines:
```lua
componentPaths.HyperLink = componentPaths.Hyperlink
componentPaths.SidebarPage = componentPaths.SideBarPage
```

i.e., specifically mention legacy names for certain components at the `componentPaths` level.

Here are some pros and cons to this approach (compared to the `string.lower` one):

Not using `string.lower` means we don't support weird capitalization that isn't officially documented, i.e. `createTEMPLaTE` or something. Each of the weird capitalizations would result in an extra function being generated by the `__index` metamethods and added to `Category`, `Template` and/or `mcm`. So it may be nice to limit for that reason. (Although I really doubt people would even think to try something like `mwse.mcm.createTEMPLaTE` unless they specifically knew they could. But it might be a bit weird that the `"create"` part needs to be lowercaes, while the `className` part doesn't.)

The downside is that this approach would theoretically be less stable than using `string.lower`, because it really depends on the capitalization of the file names. And this part makes me nervous. But I'm also not sure how much of a difference this would make, since the file names should (theoretically) always be capitalized in the same way.

File: misc/package/Data Files/MWSE/core/mcm/mcm.lua (hunk 8 above)
Comment (C3pa) [reply misc/package/Data Files/MWSE/core/mcm/mcm.lua]: I'd say your reasoning is also valid. I may have overdone it with the nesting reduction 😀.

File: misc/package/Data Files/MWSE/core/mcm/components/Component.lua [hunk 9]
```diff
@@ -112,22 +115,10 @@ function Component:getComponent(componentData)
...
-			break
-		end
-	end
-	if component then
-		--- @cast component mwseMCMComponent
-		self:prepareData(componentData)
-		return component:new(componentData)
+	local componentClass = fileUtils.getComponentClass(componentData.class)
+	if componentClass then
+		componentData.parentComponent = self
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/components/Component.lua:120]: I forgot to ask last time. This line is new. It's also added to the Category and Template. Can you explain why it is needed?

Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/components/Component.lua:120]: Good question. I also wanted to ask what you thought about this.

This line was added because I've marked `Component:prepareData` as deprecated, and previously the line 
```lua
componentData.parentComponent = self
```
would happen in `Component:prepareData`.

My reasons for marking `prepareData` as deprecated are as follows:
1. It was only used 3 times and was a very short function. However, it was used once in three separate files, which (in my opinion) made the code a bit harder to read.
2. The `__index` metamethods are already fairly complex and involve a lot of machinery spread out across several files. So I thought it would be clearer to remove the call to `prepareData` and just include the function body of `prepareData` in the `__index` metamethods themselves. (Same thing for `Component:getComponent(()`.)
3. I found the name to be a bit ambiguous (but this is more of a personal thing). In particular, the name `prepareData` suggests (to me) that this is a function that is likely to be overloaded by child classes. But every class uses the implementation specified by `mwseMCMComponent`.

What do you think about this change? Should I revert things back so that they use `prepareData`, or do you prefer the implementation without `prepareData`?

Another idea would be to make all the `prepareData` stuff (aside from `componentData.parentComponent = self`) happen in the relevant `Class:new()` methods. Since the `prepareData` stuff is currently spread out across three separate functions, and it seems like the objects themselves should have more say over how they're constructed. Since most things call `Component:new` anyways, it seems like `Component:new()` might be a nice new home for most of this logic. 

One example of this that I would like to change would be the 
```lua
if type(data) == "string" then
    if componentClass.componentType == "Template" then
        data = { name = data }
    else
        data = { label = data }
    end
end
```
code that runs inside `mwse.mcm.create<Component|Variable>`. I think it might make more sense if there were two separate `type(data) == "string"` checks that happened: one happens in `Component:new()` and the other happens in `Template:new()`.

Comment (C3pa) [reply misc/package/Data Files/MWSE/core/mcm/components/Component.lua:120]: I agree. Let's remove `Component:prepareData` altogether. Also, let's inline the code block you gave:
```lua
if type(data) == "string" then
    if componentClass.componentType == "Template" then
        data = { name = data }
    else
        data = { label = data }
    end
end
```
into Component and Template constructors.

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua (hunk 3 above)
Comment (C3pa) [reply misc/package/Data Files/MWSE/core/mcm/fileUtils.lua]: I think we are fine as is.

File: misc/package/Data Files/MWSE/core/mcm/components/Component.lua (hunk 9 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/components/Component.lua:120]: I tried to implement this and unfortunately it wasn't as straightforward as I had originally hoped. The problem is that `Component:new` and `Page:new` both would really like to access the `parentComponent` key of the `data` parameter. But moving the code block 
```lua
if type(data) == "string" then
    if componentClass.componentType == "Template" then
        data = { name = data }
    else
        data = { label = data }
    end
end
```
into the constructors means that `parentComponent = self` can only be written after calling the constructors.

I tried to play around with a few different approaches to this problem, but I think the most straightforward way is to just keep the data sanitation code in the `.__index` metamethods, even if it's a bit clunky.

I think I can simplify the code in `mcm.__index` a bit though, since `Template` already converts `data.label -> data.name` in its constructor.

Comment (C3pa) [reply misc/package/Data Files/MWSE/core/mcm/components/Component.lua:120]: Ok, let's do it. And please remove the `Component:prepareData`.

Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/components/Component.lua:120]: Will do!

Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/components/Component.lua:105]: I removed the `prepareData` method and tried to tidy up the constructors as best I could.

I also removed
```lua
function Component:__index(key)
    return self[key]
end
```
for the following reasons:
- The first time `Component:new()` is called, it would set `Component.__index = Component`, ignoring the above definition.
- The above definition would result in a stack overflow if it was ever executed. This is because `__index(self, key)` is only ever called if `key` is not in `self`.

In addition, I also tidied up the `fileUtils.lua` file so that the components are more descriptive and so that variables are initialized closer to where they are first used.

I know you warned about making unnecessary code changes, but I wasn't sure about how much that applied to `fileUtils` since it's being created in this PR. I tried to keep most of the changes to a minimum though. (Or at the very least, to make them easier to follow.)

One change I'm a bit on the fence about is changing 
```lua
local componentPaths = {}
local variablePaths = {}
```
to 
```lua
local componentPaths, variablePaths = {}, {}
```
In general, I don't love the `local a, b = val1, val2` syntax, but I did it here for IDE reasons:
- When the variables were being defined on different lines, the comment would only appear in the descriptive popup for `componentPaths` (i.e., the popup that gets shown when hovering your mouse over a local variable).
- Defining both variables in one expression means this popup now gets applied to both local variables.

I'm not really sure if this is worth it, and I would be happy to walk back this change if you would prefer.

File: misc/package/Data Files/MWSE/core/mcm/fileUtils.lua [hunk 10]
```diff
@@ -13,8 +13,9 @@ local variables = {}  ---@type table<string, mwseMCMVariable>
 -- Files get `require`d in by the `fileUtils.get<Component|Variable>Class` functions. (They actually get `include`d for stability reasons.)
 -- Once a file has been `require`d by the `fileUtils.get<Component|Variable>Class` functions, its path will be removed from these tables.
 -- (Testing has shown that this approach results in a noticeable boost in performance.)
----@type table<string, string>, table<string, string>
-local componentPaths, variablePaths = {}, {}
+
+local componentPaths = {} ---@type table<string, string>
```
[the remark this answers is not in the document]
Comment (C3pa) [reply misc/package/Data Files/MWSE/core/mcm/fileUtils.lua:17]: Both are okay.

File: misc/package/Data Files/MWSE/core/mcm/components/Component.lua (hunk 9 above)
Comment (C3pa) [reply misc/package/Data Files/MWSE/core/mcm/components/Component.lua:105]: I am fine with both `local a, b = val1, val2` and each on a separate line.

About the unnecessary changes, if you've got to change something, change it. It's mostly from the perspective of the reviewer, that it's easier to understand the implications of changes when the diff is smaller and when things are split into logical commits. For example, you can have a look at #440. When you look at the diff in the PR, it's quite hard to see what's all that going on there. It makes much more sense when you inspect it's diff commit by commit. (A disclaimer, as the author of the given PR, I don't imply that it's the best/perfect example of submitting code for review, I just hope I illustrated the point.). 

In conclusion, you save me some time when the diff is smaller, but if you've got to change something go ahead.

**Review by C3pa — APPROVED** [review]
Thanks Herbert
